src/charm.py

```diff
@@ -110,63 +102,191 @@ def _missing_mandatory_relations(self) -> Optional[str]:
             return
+
         if not self._container.can_connect():
-            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.add_status(WaitingStatus("Waiting for container to be ready"))
             return
+
         if not self._database_is_available():
-            self.unit.status = WaitingStatus("Waiting for `database` relation to be available")
+            event.add_status(WaitingStatus("Waiting for `database` relation to be available"))
             return
+
         if not self._nrf_is_available():
-            self.unit.status = WaitingStatus("Waiting for NRF relation to be available")
+            event.add_status(WaitingStatus("Waiting for NRF relation to be available"))
             return
+
         if not self._storage_is_attached():
-            self.unit.status = WaitingStatus("Waiting for storage to be attached")
+            event.add_status(WaitingStatus("Waiting for storage to be attached"))
+            return
+
+        if not _get_pod_ip():
+            event.add_status(WaitingStatus("Waiting for pod IP address to be available"))
+            return
+
+        if self._csr_is_stored() and not self._get_current_provider_certificate():
+            event.add_status(WaitingStatus("Waiting for certificates to be stored"))
+            return
+
+        if not self._smf_service_is_running():
+            event.add_status(WaitingStatus("Waiting for SMF service to start"))
             return
+
+        event.add_status(ActiveStatus())
+
+    def _smf_service_is_running(self) -> bool:
+        """Check if the SMF service is running.
+
+        Returns:
+            bool: Whether the SMF service is running.
+        """
+        if not self._container.can_connect():
+            return False
+        try:
+            service = self._container.get_service(self._service_name)
+        except ModelError:
+            return False
+        return service.is_running()
+
+    def ready_to_configure(self) -> bool:
+        """Returns whether the preconditions are met to proceed with the configuration.
+
+        Returns:
+            ready_to_configure: True if all conditions are met else False
+        """
+        if self._missing_mandatory_relations():
+            return False
+
+        if not self._container.can_connect():
+            return False
+
+        if not self._database_is_available():
+            return False
+
+        if not self._nrf_is_available():
+            return False
+
+        if not self._storage_is_attached():
+            return False
+
         if not _get_pod_ip():
-            self.unit.status = WaitingStatus("Waiting for pod IP address to be available")
+            return False
+
+        return True
+
+    def _configure_sdcore_smf(self, event: EventBase) -> None:  # noqa C901
+        """Adds pebble layer and manages Juju unit status.
+
+        Args:
+            event: Juju event
+        """
+        if not self.ready_to_configure():
```

Comment:
Should we log the fact the the workload wasn't ready to be configured?

Reply:
Yes, we can log it.

Reply:
Done